### PR TITLE
[Access] 4506 - Make CCF encoded events optional via request parameter

### PR DESCRIPTION
**Pull request by Guitarheroua**
https://github.com/onflow/flow-go/issues/4506

### Context

This pull request introduces a new request parameter for a list of endpoints from the [AccessAPI](https://github.com/onflow/flow-go/blob/master/access/api.go) and the [ExecutionDataAPI](https://github.com/onflow/flow-go/blob/dafd8fd29c476cacc497f79940412ba377f723fe/engine/access/state_stream/backend.go#L73), which specifies the encoding for events. If not specified, default encoding remains JSON.

As part of this PR:

* The protobuf structures have been extended (see https://github.com/onflow/flow/pull/1391)
* flow-emulator PR (https://github.com/onflow/flow-emulator/pull/484)

### CODE REVIEW

File: access/legacy/handler.go [hunk 1]
```diff
@@ -189,7 +189,7 @@ func (h *Handler) GetTransactionResult(
 ) (*accessproto.TransactionResultResponse, error) {
 	id := convert.MessageToIdentifier(req.GetId())
 
-	result, err := h.api.GetTransactionResult(ctx, id, flow.ZeroID, flow.ZeroID)
+	result, err := h.api.GetTransactionResult(ctx, id, flow.ZeroID, flow.ZeroID, nil)
```
Comment (peterargue) [inline access/legacy/handler.go:192]: these should default to json-cdc in the legacy handler

File: engine/access/rest/routes/events.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ func GetEvents(r *request.Request, backend access.API, _ models.LinkGenerator) (
 	// if the request has block IDs provided then return events for block IDs
 	var blocksEvents models.BlocksEvents
 	if len(req.BlockIDs) > 0 {
-		events, err := backend.GetEventsForBlockIDs(r.Context(), req.Type, req.BlockIDs)
+		events, err := backend.GetEventsForBlockIDs(r.Context(), req.Type, req.BlockIDs, nil)
```
Comment (peterargue) [inline engine/access/rest/routes/events.go:24]: rest as well

File: engine/access/rpc/backend/backend_transactions.go [hunk 3]
```diff
@@ -416,6 +418,8 @@ func (b *backendTransactions) GetTransactionResultsByBlockID(
 		"number of transaction results returned by execution node is less than the number of transactions  in the block",
 	)
 
+	eventEncodingVersion := convert.GetConversionEventEncodingVersion(eventEncodingVersionValue)
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_transactions.go:421]: same comment about verifying the from/to conversions for the tx result endpoints

File: engine/common/rpc/convert/events.go [hunk 4]
```diff
@@ -258,3 +261,26 @@ func BlockEventsToMessage(block flow.BlockEvents) (*accessproto.EventsResponse_R
...
 	}, nil
 }
+
+// GetConversionEventEncodingVersion returns the appropriate event encoding version for conversion.
+// Execution node events payloads are always encoded in CCF. If eventEncodingVersionValue is JSON-CDC,
+// CFF-encoded payloads should be converted to JSON-CDC, and CFF is returned as the format to convert from.
+// If eventEncodingVersionValue is CFF, payloads will not be converted and will be left as is. In this case,
+// JSON-CDC is returned as the format to convert from, which will be ignored in converters, and payloads
+// will be returned as-is.
+func GetConversionEventEncodingVersion(eventEncodingVersionValue *entities.EventEncodingVersionValue) entities.EventEncodingVersion {
```
Comment (peterargue) [inline engine/common/rpc/convert/events.go:271]: this is a little confusing. What do you think about switching to use `entities.EventEncodingVersion` everywhere and explicitly specify the default encoding.

**Review by peterargue — COMMENTED** [review]
thanks for tackling this @Guitarheroua!

A little context on the existing code. We originally added support for CCF encoding in somewhat of a hurry while rolling out the feature. It ended up with a bunch of awkward interfaces and tech debt.

You're starting to unravel it here, but I think it would be good to flip the approach a bit.

Currently, event data is conceptually pulled through access nodes, from execution nodes. In this model, ENs set the encoding, and the events are updated from that original value to JSON-CDC. This made sense during the transition, but gets a little messy especially as we're moving to storing events locally on ANs.

I think a more useful way to think about it is that events have a native encoding that is converted to the requested version as needed. In the current software, ENs and ExecutionData exclusively use CCF. This PR is updating to allow clients to request a specific version. Now instead of converting from some version, we're converting to some version.

There isn't really much to change in this PR besides from terminology, but hopefully this is helpful context.

There's also a change coming to support reading events from the local db. In this case, events will be assumed to be CCF encoded.

File: engine/common/rpc/convert/execution_data.go [hunk 5]
```diff
@@ -25,6 +31,7 @@ func BlockExecutionDataEventPayloadsToJson(m *entities.BlockExecutionData) error
 			m.ChunkExecutionData[i].Events[j].Payload = converted
 		}
 	}
+
```
Comment (durkmurder) [inline engine/common/rpc/convert/execution_data.go:34]:
```suggestion
```

File: engine/access/state_stream/handler_test.go [hunk 6]
```diff
@@ -110,6 +112,114 @@ func TestExecutionDataStream(t *testing.T) {
...
+
+	// Test scenario for JSON event encoding.
+	t.Run("test JSON event encoding", func(t *testing.T) {
+		makeStreamRequest(&executiondata.SubscribeExecutionDataRequest{
+			EventEncodingVersion: entities.EventEncodingVersion_JSON_CDC_V0,
+		})
+		for {
+			resp, err := stream.RecvToClient()
+			if err == io.EOF {
+				break
```
Comment (durkmurder) [inline engine/access/state_stream/handler_test.go:179]: how do we hit this condition? If test just breakes here it's still a success even though no entries were actually asserted.

File: engine/access/rpc/backend/backend_test.go [hunk 7]
```diff
@@ -636,7 +638,7 @@ func (suite *Suite) TestTransactionStatusTransition() {
 		Times(len(fixedENIDs)) // should call each EN once
 
 	// first call - when block under test is greater height than the sealed head, but execution node does not know about Tx
-	result, err := backend.GetTransactionResult(ctx, txID, flow.ZeroID, flow.ZeroID)
+	result, err := backend.GetTransactionResult(ctx, txID, flow.ZeroID, flow.ZeroID, entitiesproto.EventEncodingVersion_JSON_CDC_V0)
```
Comment (durkmurder) [inline engine/access/rpc/backend/backend_test.go:641]: `EventEncodingVersion_JSON_CDC_V0` is used all over the place as default value, is it possible to introduce a variable for it? Something like: `DefaultEventEncodingVersion`?

File: engine/access/rpc/backend/backend_test.go [hunk 8]
```diff
@@ -1097,11 +1107,33 @@ func (suite *Suite) TestGetEventsForBlockIDs() {
...
+			params.ExecutionReceipts = receipts
+			params.ConnFactory = connFactory
+			params.FixedExecutionNodeIDs = validENIDs.Strings()
+
+			// create the handler
+			backend, err := New(params)
+			suite.Require().NoError(err)
+
+			// execute request with an empty block id list and expect an empty list of events and no error
+			result, err := backend.GetEventsForBlockIDs(ctx, string(flow.EventAccountCreated), []flow.Identifier{}, entitiesproto.EventEncodingVersion_JSON_CDC_V0)
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_test.go:1127]: should this be? 
```suggestion
			result, err := backend.GetEventsForBlockIDs(ctx, string(flow.EventAccountCreated), []flow.Identifier{}, version)
```

File: engine/access/state_stream/backend/handler_test.go [hunk 9]
```diff
@@ -197,6 +199,7 @@ func (s *HandlerTestSuite) TestHeartbeatResponse() {
 	})
 }
 
+// TestExecutionDataStream tests the execution data stream with different event encoding versions.
 func TestExecutionDataStream(t *testing.T) {
```
Comment (peterargue) [inline engine/access/state_stream/backend/handler_test.go:203]: Can you add a test covering `GetExecutionDataByBlockID`?

It looks like we don't have any existing set for that endpoint, so only happy path tests covering both encodings is needed.

File: engine/access/state_stream/backend/handler_test.go [hunk 10]
```diff
@@ -315,21 +361,15 @@ func TestEventStream(t *testing.T) {
...
-
 	h := NewHandler(api, flow.Localnet.Chain(), config)
 
 	wg := sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
 		wg.Done()
-		err := h.SubscribeEvents(&executiondata.SubscribeEventsRequest{}, stream)
+		err := h.SubscribeEvents(&executiondata.SubscribeEventsRequest{
+			EventEncodingVersion: entities.EventEncodingVersion_JSON_CDC_V0,
```
Comment (peterargue) [inline engine/access/state_stream/backend/handler_test.go:371]: can you also include a test for `SubscribeEvents` with CCF

**Review by peterargue — APPROVED** [review]
looks good. Thanks @Guitarheroua!

File: engine/common/rpc/convert/execution_data_test.go [hunk 11]
```diff
@@ -3,6 +3,8 @@ package convert_test
 import (
 	"testing"
 
+	"github.com/onflow/flow/protobuf/go/flow/entities"
```
Comment (durkmurder) [inline engine/common/rpc/convert/execution_data_test.go:6]: can we group it with other `github.com/onflow` imports?

File: utils/unittest/generator/events.go [hunk 12]
```diff
@@ -107,3 +97,15 @@ func (g *Events) New() flow.Event {
 
 	return event
 }
+
+// GetEventsWithEncoding generates a specified number of events with a given encoding version.
+func GetEventsWithEncoding(n int, version entities.EventEncodingVersion) []flow.Event {
+	eventGenerator := EventGenerator(WithEncoding(version))
+
```
Comment (durkmurder) [inline utils/unittest/generator/events.go:104]:
```suggestion
```

File: utils/unittest/generator/events.go [hunk 13]
```diff
@@ -107,3 +97,15 @@ func (g *Events) New() flow.Event {
...
+	events := make([]flow.Event, 0, n)
+	for i := 0; i < n; i++ {
+		events = append(events, eventGenerator.New())
+	}
+
```
Comment (durkmurder) [inline utils/unittest/generator/events.go:109]:
```suggestion
```

File: engine/common/rpc/convert/events_test.go [hunk 14]
```diff
@@ -3,13 +3,14 @@ package convert_test
 import (
 	"testing"
 
+	"github.com/onflow/flow/protobuf/go/flow/entities"
```
Comment (durkmurder) [inline engine/common/rpc/convert/events_test.go:6]: Please group it with other `github.com/onflow` imports

File: engine/access/state_stream/backend/handler_test.go [hunk 15]
```diff
@@ -15,6 +15,9 @@ import (
 	pb "google.golang.org/genproto/googleapis/bytestream"
 	"google.golang.org/grpc"
 
+	"github.com/onflow/cadence/encoding/ccf"
```
Comment (durkmurder) [inline engine/access/state_stream/backend/handler_test.go:18]: Please group it with other `github.com/onflow` imports ?

File: engine/access/rpc/backend/retry_test.go [hunk 16]
```diff
@@ -3,6 +3,8 @@ package backend
 import (
 	"context"
 
+	"github.com/onflow/flow/protobuf/go/flow/entities"
```
Comment (durkmurder) [inline engine/access/rpc/backend/retry_test.go:6]: Please group it with other `github.com/onflow` imports ?

File: engine/access/state_stream/backend/handler_test.go [hunk 17]
```diff
@@ -197,186 +191,333 @@ func (s *HandlerTestSuite) TestHeartbeatResponse() {
...
+
+			response, err := h.GetExecutionDataByBlockID(ctx, &executiondata.GetExecutionDataByBlockIDRequest{
+				BlockId:              blockID[:],
+				EventEncodingVersion: test.eventVersion,
+			})
+			require.NoError(t, err)
+			require.NotNil(t, response)
+
+			blockExecutionData := response.GetBlockExecutionData()
+
```
Comment (durkmurder) [inline engine/access/state_stream/backend/handler_test.go:239]:
```suggestion
```

**Review by durkmurder — APPROVED** [review]

**Pull request merged**